### PR TITLE
Add SameSite directive #174

### PR DESCRIPTION
**Pull request by MadCat34**
This PR allow to add the SameSite directive in SetCookie header

See #174

### CODE REVIEW

**Comment by MadCat34** [comment]
Hi @webimpress 
Thanks for your return.

I have added a constant for None value.

File: src/Header/SetCookie.php [hunk 1]
```diff
@@ -17,6 +17,29 @@
...
+
+    /**
+     * Cookie will not be passed for any cross-domain requests unless it's a regular link that navigates user
+     * to the target site.
+     * Other requests methods (such as POST and PUT) and XHR requests will not contain this cookie.
+     */
+    const SAME_SITE_LAX = 'Lax';
+
+    /**
+     * Vookie will be sent with same-site and cross-site requests.
```
Comment (webimpress) [inline src/Header/SetCookie.php]: Typo: cookie

File: src/Header/SetCookie.php [hunk 2]
```diff
@@ -17,6 +17,29 @@
...
+     */
+    const SAME_SITE_NONE = 'None';
+
+    const SAME_SITE_ALLOWED_VALUES = [
+        self::SAME_SITE_STRICT,
+        self::SAME_SITE_LAX
```
Comment (webimpress) [inline src/Header/SetCookie.php]: Should contain also NONE

File: src/Header/SetCookie.php [hunk 3]
```diff
@@ -572,6 +611,24 @@ public function isSessionCookie()
...
+    /**
+     * @return string|null
+     */
+    public function getSameSite()
+    {
+        return $this->sameSite;
+    }
+
+    /**
+     * @param string $sameSite
```
Comment (webimpress) [inline src/Header/SetCookie.php]: string or null

File: src/Header/SetCookie.php [hunk 4]
```diff
@@ -572,6 +611,24 @@ public function isSessionCookie()
...
+     * @return SetCookie
+     */
+    public function setSameSite($sameSite)
+    {
+        $this->sameSite = $sameSite;
```
Comment (webimpress) [inline src/Header/SetCookie.php]: I believe we should check here if we set the correct value, and throw exception if not?

File: src/Header/SetCookie.php [hunk 5]
```diff
@@ -199,6 +230,7 @@ public static function fromString($headerLine, $bypassHeaderFieldName = false)
      * @param   bool                $httponly
      * @param   string              $maxAge
      * @param   int                 $version
+     * @param   string              $sameSite
```
Comment (webimpress) [inline src/Header/SetCookie.php]: string|null, I see other types should be updated also, but this is out of scope

File: src/Header/SetCookie.php [hunk 6]
```diff
@@ -298,6 +333,11 @@ public function getFieldValue()
             $fieldValue .= '; HttpOnly';
         }
 
+        $sameSite = $this->getSameSite();
+        if ($sameSite !== null && in_array($sameSite, self::SAME_SITE_ALLOWED_VALUES, true)) {
```
Comment (webimpress) [inline src/Header/SetCookie.php]: first condition is redundant, `in_array` is sufficient

File: src/Header/SetCookie.php (hunk 2 above)
Comment (MadCat34) [reply src/Header/SetCookie.php]: Done !

File: src/Header/SetCookie.php (hunk 4 above)
Comment (MadCat34) [reply src/Header/SetCookie.php]: Sure :-)

File: src/Header/SetCookie.php [hunk 7]
```diff
@@ -572,6 +612,30 @@ public function isSessionCookie()
...
+     * @param string|null $sameSite
+     * @return SetCookie
+     * @throws Exception\InvalidArgumentException
+     */
+    public function setSameSite($sameSite)
+    {
+        if (! in_array($sameSite, self::SAME_SITE_ALLOWED_VALUES, true)) {
+            throw new Exception\InvalidArgumentException(
+                '"' . $sameSite . '" is not an allowed value (Strict, Lax, None) for SameSite directive.'
+            );
```
Comment (webimpress) [inline src/Header/SetCookie.php]: Need unit test for that case, and we are good :) thx

File: src/Header/SetCookie.php (hunk 6 above)
Comment (MadCat34) [inline src/Header/SetCookie.php]: $sameSite !== null must be tested because if $sameSite is null, we don't want `; SameSite=` in header and null must not be in allowed value (it breaks later)

File: src/Header/SetCookie.php (hunk 7 above)
Comment (MadCat34) [reply src/Header/SetCookie.php]: ok :)

File: src/Header/SetCookie.php (hunk 6 above)
Comment (webimpress) [reply src/Header/SetCookie.php]: yeah, and as we can't set invalid value we don't need in_array check ;-)

File: src/Header/SetCookie.php (hunk 7 above)
Comment (MadCat34) [reply src/Header/SetCookie.php]: Done :-)

**Comment by michalbundyra** [comment]
Thanks, @MadCat34!
